walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,96 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  
+  private Collection<TimeRange> timeOptions = new ArrayList<TimeRange>();
+  
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    timeOptions.clear();
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    
+    if (request.getAttendees().isEmpty() || events.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    
+    Set<String> allEventAttendees = new HashSet<>();
+    for (Event event : events) {
+      allEventAttendees.addAll(event.getAttendees());
+    }
+    Set<String> allRequestAttendees = new HashSet<>();
+    allRequestAttendees.addAll(request.getAttendees());
+    for (String attendee : allRequestAttendees) {
+      if (!allEventAttendees.contains(attendee)) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+    }
+    
+    List<Event> eventsList = new ArrayList<Event>();
+    eventsList.addAll(events);
+    Collections.sort(eventsList, Event.ORDER_BY_START);
+    TimeRange firstEventTime = eventsList.get(0).getWhen();
+    TimeRange lastEventTime = eventsList.get(eventsList.size() - 1).getWhen();
+    boolean earlyEnd = false;
```

Comment:
Is this `notAllEventsProcessed`?

If so, that's more self-explanatory at the usage sites than "earlyEnd" which just tells me something "ended" "early".

Reply:
Done